genmethod.go

```diff
@@ -179,9 +179,12 @@ func (m *marshalMethod) receiver() Receiver {
 	return r
 }
 
-func (m *marshalMethod) intermediateType(name string) Struct {
+func (m *marshalMethod) intermediateType(name string, forMarshal bool) Struct {
 	s := Struct{Name: name}
 	for _, f := range m.mtyp.Fields {
+		if !forMarshal && f.function != nil {
```

Comment:
You can use marshalMethod.isUnmarshal instead of passing the forMarshal flag as a parameter.